web/modules/custom/dpl_react_apps/dpl_react_apps.module

```diff
@@ -114,3 +130,48 @@ function dpl_react_apps_texts_renewal(): array {
     'renew-processing-text' => t('Processing...', [], ['context' => 'Renew loan']),
   ];
 }
+
+/**
+ * Make sure that generated url is a string.
+ *
+ * @param string|\Drupal\Core\GeneratedUrl $url
+ *   Drupal generated Url object.
+ */
+function dpl_react_apps_ensure_url_is_string(string|GeneratedUrl $url): string {
+  if ($url instanceof GeneratedUrl) {
+    $url = $url->getGeneratedUrl();
+  }
+
+  return $url;
+}
+
+/**
+ * Get the base url of the dpl cms.
+ *
+ * @return string
+ *   The base url of the dpl cms.
+ */
+function dpl_react_apps_get_dpl_cms_base_url(): string {
+  return rtrim(dpl_react_apps_ensure_url_is_string(
+    Url::fromRoute('<front>', [], ['absolute' => TRUE])->toString()
+  ), "/");
+}
+
+/**
+ * Format urls used by the apps.
+ *
+ * @param string|null $url
+ *   Url to format.
+ * @param string $fallbackUrl
+ *   If the url param is not present.
+ *
+ * @return string
+ *   Formatted url.
+ */
+function dpl_react_apps_format_app_url(string|NULL $url, string $fallbackUrl = ''): string {
+  if (!empty($url)) {
+    return $url;
+  }
+
+  return $fallbackUrl ?: dpl_react_apps_get_dpl_cms_base_url();
```

Comment:
Consider adding a comment about why we think it is a good idea to fall back to the base url if no url is provided.